### PR TITLE
IDS-536: update ingestion scripts with interim storagebox changes

### PR DESCRIPTION
**Pull request by uoa-noel**
This PR implements an interm solution for file transfer and a convention for storing datafiles. It was decided that ingestion scripting would not be responsible for replicating the datafiles to multiple StorageBoxes. Instead, the ingestion script would transfer files into one _staging StorageBox_ which will be filesystem-based for the pipeline. Later, as part of server-side validation, functionality will be added to MyTardis to create _project-specific StorageBoxes_, and replicate datafiles to different boxes.

Major changes in this PR are:
- Change Conveyor to remove unnecessary abstractions, as we are only transferring to one StorageBox. Filesystem-based transfer using rsync is supported.
- Add an interm convention for file storage in the _staging StorageBox_. Each file will be stored in `[storagebox path]/ds-[dataset id]/[directory]/[filename]`
- Removal of project-specific StorageBox code. This code is saved on the `wip-project-specific-storagebox` branch for future reference.
- Conveyor is now run as part of the ingest_datafiles() method in `IngestionFactory`.

### CODE REVIEW

**Review by andrew-uoa — COMMENTED** [review]
Nice work, I think the storage code structure is much clearer and more understandable now, and there's less code, which is always welcome.

I just raised a few small/isolated issues/suggestions - happy to discuss them

File: src/ingestion.py [hunk 1]
```diff
@@ -50,8 +55,24 @@ def main(
...
+    except ValidationError as error:
+        logger.error(
+            (
+                "An error occurred while validating the environment "
+                "configuration. Make sure all required variables are set "
+                "or pass your own configuration instance. Error: %s"
+            ),
+            error,
+        )
+        sys.exit()
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/ingestion.py:75]: Good catch!

File: src/conveyor/conveyor.py [hunk 2]
```diff
@@ -1,60 +1,119 @@
...
+    """A custom exception for transfer failures."""
+
+
+def is_rsync_on_path() -> bool:
+    """Tests whether rsync is available on the PATH.
+
+    Returns:
+        bool: True if rsync is found, false if not.
+    """
+    state = subprocess.run(["rsync", "-h"], capture_output=True, check=False)  # nosec
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/conveyor/conveyor.py:25]: Good tip!

[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/conveyor/conveyor.py]: It seems there isn't a big difference between the two approaches, and with a `subprocess.run`, it's a check that ensures the command is executable, so I will stay with the current approach if that's alright.

File: samples/ro_crate_ingestion_example.py [hunk 3]
```diff
@@ -70,12 +68,6 @@ def ingest_data(ro_crate_path: Path, rsync_pth: Path) -> None:
     factory.ingest_datafiles(metadata.get_datafiles())
 
     datafiles = metadata.get_datafiles()
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply samples/ro_crate_ingestion_example.py]: Removed.

File: src/conveyor/conveyor.py [hunk 4]
```diff
@@ -1,60 +1,119 @@
...
-        """Initialises the file ingestion object.
+    def __init__(self, store: FilesystemStorageBoxConfig) -> None:
+        """Initialises the file transfer Conveyor object.
 
         Args:
             transport (AbstractTransport): The concrete transport mechanism to use.
         """
-        self.transport = transport
+        if not is_rsync_on_path():
+            logger.error("Could not find rsync on PATH.")
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/conveyor/conveyor.py]: Have now added a `RuntimeException` if `rsync` is not found.

File: src/conveyor/conveyor.py [hunk 5]
```diff
@@ -1,60 +1,119 @@
...
+                list_f.write(str(path) + "\n")
+            list_f.flush()
+            # Run rsync commandline to transfer the files.
+            # Disable bandit's warning about subprocess.
+            result = subprocess.run(  # nosec
+                ["rsync", "-av", "--files-from", list_f.name, src, destination_dir],
+                check=False,
+            )
+            if result.returncode > 0:
+                raise FailedTransferException("rsync return code was not 0.")
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/conveyor/conveyor.py:98]: Good point, now we log the rsync output. Note due to piping, rsync output will not be visible at the commandline until after rsync finishes. Have added logging message explaining this.

File: src/ingestion_factory/factory.py [hunk 6]
```diff
@@ -316,6 +306,17 @@ def ingest_datafiles(
...
+        # Create a file transfer with the conveyor
+        logger.info("Starting transfer of datafiles.")
+        try:
+            self.conveyor.transfer(self.config.source_directory, prepared_datafiles)
+            logger.info("Finished transferring datafiles.")
+        except FailedTransferException:
+            logger.error(
+                "Datafile transfer could not complete. Check rsync output for more information."
+            )
+            sys.exit()
```
[the remark this answers is not in the document]
Comment (uoa-noel) [reply src/ingestion_factory/factory.py]: Removed sys.exit() call.

**Review by andrew-uoa — APPROVED** [review]
Thanks for the changes, looks good to me!

**Pull request merged**